<?xml version="1.0"?>
<launch>
  <!-- ROS paramaters -->
  <arg name="bagfile" default=""/>
  <arg name="topic" default=""/>
  <arg name="odom_frame" default="odom"/>
  <arg name="base_frame" default=""/>
  <arg name="publish_odom_tf" default="true"/>
  <arg name="visualize" default="true"/>

  <!-- damm-loam parameter sets -->
  <arg name="config_file" default=""/>
  <arg name="params_file" default="params.yaml"/>
  <arg name="deskew" default="false"/>
  <arg name="max_range" default="100.0"/>
  <arg name="min_range" default="0.3"/>
  <arg name="voxel_size" default="0.4"/>
  <arg name="map_cleanup_radius" default="1500"/>
  <arg name="desired_num_voxelized_points" default="1500"/>
  <arg name="planarity_threshold" default="0.2"/>
  <arg name="max_points_per_voxel" default="1"/>
  <arg name="initial_threshold" default="2.0"/>
  <arg name="min_motion_th" default="0.1"/>
  <arg name="max_num_iterations" default="100"/>
  <arg name="convergence_criterion" default="0.0001"/>

  <!-- Load parameters from params.yaml -->
  <rosparam command="load" file="$(find damm_loam)/config/$(arg params_file)"/>

  <!-- Odometry node -->
  <node pkg="damm_loam" type="odometry_node" name="odometry_node" output="screen">
    <!-- ROS params -->
    <remap from="pointcloud_topic" to="$(arg topic)"/>
    <param name="odom_frame" value="$(arg odom_frame)"/>
    <param name="base_frame" value="$(arg base_frame)"/>
    <param name="publish_odom_tf" value="$(arg publish_odom_tf)"/>
    <param name="visualize" value="$(arg visualize)"/>
    
    <!-- Conditional Parameter Loading -->
    <!-- Load parameters from config file if provided -->
    <rosparam if="$(eval config_file != '')" command="load" file="$(find damm_loam)/config/$(arg config_file)"/>
    
    <!-- Load parameters directly if config_file is empty -->
    <param if="$(eval config_file == '')" name="deskew" value="$(arg deskew)"/>
    <param if="$(eval config_file == '')" name="max_range" value="$(arg max_range)"/>
    <param if="$(eval config_file == '')" name="min_range" value="$(arg min_range)"/>
    <param if="$(eval config_file == '')" name="voxel_size" value="$(arg voxel_size)"/>
    <param if="$(eval config_file == '')" name="map_cleanup_radius" value="$(arg map_cleanup_radius)"/>
    <param if="$(eval config_file == '')" name="desired_num_voxelized_points" value="$(arg desired_num_voxelized_points)"/>
    <param if="$(eval config_file == '')" name="planarity_threshold" value="$(arg planarity_threshold)"/>
    <param if="$(eval config_file == '')" name="max_points_per_voxel" value="$(arg max_points_per_voxel)"/>
    <param if="$(eval config_file == '')" name="initial_threshold" value="$(arg initial_threshold)"/>
    <param if="$(eval config_file == '')" name="min_motion_th" value="$(arg min_motion_th)"/>
    <param if="$(eval config_file == '')" name="max_num_iterations" value="$(arg max_num_iterations)"/>
    <param if="$(eval config_file == '')" name="convergence_criterion" value="$(arg convergence_criterion)"/>
  </node>

  <!-- LiDAR Projection Node -->
  <node pkg="damm_loam" type="lidar_projection_node" name="lidar_projection_node" output="screen"/>

  <!-- Normal Extraction Node -->
  <node pkg="damm_loam" type="normal_extraction_node" name="normal_extraction_node" output="screen"/>

  <!-- Rosbag player, if bagfile argument was specified -->
  <group if="$(eval bagfile != '')">
    <node pkg="rosbag" type="play" name="player" args="$(arg bagfile)" output="screen"/>
  </group>

  <!-- Visualizer -->
  <group if="$(arg visualize)">
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find damm_loam)/rviz/damm_loam_ros1.rviz"/>
  </group>

</launch>
